hamilton/stores/base.py

```diff
@@ -0,0 +1,246 @@
+import abc
+import pickle
+from datetime import datetime, timedelta, timezone
+from typing import Any, Optional, Sequence, Tuple, Type
+
+from hamilton.htypes import custom_subclass_check
+from hamilton.io.data_adapters import DataLoader, DataSaver
+from hamilton.registry import LOADER_REGISTRY, SAVER_REGISTRY
+
+
+class MaterializationError(Exception):
+    pass
+
+
+class ResultRetrievalError(Exception):
+    pass
+
+
+class MetadataStoreIndexingError(Exception):
+    """Internal error that should be caught during testing."""
+
+
+# def search_data_adapter_registry(name: str, type_: type) -> Tuple[Type[DataSaver], Type[DataLoader]]:
+#     """Find pair of DataSaver and DataLoader registered with `name` and supporting `type_`"""
+#     if name not in SAVER_REGISTRY or name not in LOADER_REGISTRY:
+#         raise KeyError(f"{name} isn't associated to both a DataLoader and a DataSaver")
+
+#     data_saver_cls = None
+#     for saver_cls in SAVER_REGISTRY[name]:
+#         if saver_cls.applies_to(type_):
+#             data_saver_cls = saver_cls
+#             break
+
+#     if data_saver_cls is None:
+#         raise KeyError(f"{name} doesn't have any DataSaver supporting type {type_}")
+
+#     data_loader_cls = None
+#     for loader_cls in LOADER_REGISTRY[name]:
+#         if loader_cls.applies_to(type_):
+#             data_loader_cls = loader_cls
+#             break
+
+#     if data_loader_cls is None:
+#         raise KeyError(f"{name} doesn't have any DataLoader supporting type {type_}")
+
+#     return saver_cls, loader_cls
+
+
+def search_data_adapter_registry(
+    name: str, type_: type
+) -> Tuple[Type[DataSaver], Type[DataLoader]]:
+    """Find pair of DataSaver and DataLoader registered with `name` and supporting `type_`"""
+    if name not in SAVER_REGISTRY or name not in LOADER_REGISTRY:
+        raise KeyError(f"{name} isn't associated to both a DataLoader and a DataSaver")
+
+    # TODO edit logic to
+    try:
+        saver_cls = next(
+            saver_cls
+            for saver_cls in SAVER_REGISTRY[name]
+            if any(
+                custom_subclass_check(type_, applicable_type)
+                for applicable_type in saver_cls.applicable_types()
+            )
+        )
+    except StopIteration as e:
+        raise KeyError(f"{name} doesn't have any DataSaver supporting type {type_}") from e
+
+    try:
+        loader_cls = next(
+            loader_cls
+            for loader_cls in LOADER_REGISTRY[name]
+            if any(
+                custom_subclass_check(type_, applicable_type)
+                for applicable_type in loader_cls.applicable_types()
+            )
+        )
+    except StopIteration as e:
+        raise KeyError(f"{name} doesn't have any DataLoader supporting type {type_}") from e
+
+    return saver_cls, loader_cls
+
+
+class ResultStore(abc.ABC):
```

Comment:
I think we should move this to caching, it's not top-level (as nobody else uses it). It's purely a tool for caching -- we can generalize later if we want.